### DISCUSSION TITLE
working with multitemporal data

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lidR#601] [Q&A]
I wonder what are the possibilities in lidR to work with point clouds from more than one point in time? Are there examples?

Recently I received ALS data of a larger area where there was data from a couple of years ago for the first time. And revisited surveys are getting more and more I guess. 

My first thought was that I could maybe load both data sets in the same LAScatalog and work with it, but I got `Be careful, some tiles seem to overlap each other. lidR may return incorrect outputs with edge artifacts when processing this catalog.` which brought me to this discussion (https://gis.stackexchange.com/questions/345919/overlapping-las-tiles). So it seems working with multiple point clouds for the same area might not be intended in lidR and I guess is also not easy in general as point clouds may vary between ettributes etc.  Anyway I think there is a lot of potential to work with mutliteporal point clouds. 

Possible use cases I was thinking of include:
- multitemporal change detection, either
  - directly on the point cloud (as already dicussed for single LAS files in https://github.com/r-lidar/lidR/issues/305)
  - based on intermediate products such as e.g. the difference of CHM from points of time A and points of time B
- create metrics based on the densified point cloud of A and B together where only subtle changes are expected (e.g. terrain)

For single LAS files this should be possible I guess, but what about larger areas in a LAScatalog?

**Comment by Jean-Romain** [comment]
> create metrics based on the densified point cloud of A and B together where only subtle changes are expected (e.g. terrain)

You can read all the files as a single collection. You have a message for being careful, but it is only a message to trigger your attention. Process the collection with a chunk size != 0 and it will work. However pay attention to your data quality. It is unlikely that both dataset are perfectly aligned especially in Z.

> based on intermediate products such as e.g. the difference of CHM from points of time A and points of time B

You can do that by splitting your files in two independent groups and make two CHMs

However there is no way to have two LAScatalog and generate the same chunks for both and load two LAS at the same extents to compare the two collection chunk by chunk. That might be the object of a feature request. I can imagine something like

```r
multi_ctg = readMultiLAScatalog("folder1", "folder2")
opt_chunk_size(multi_ctg) = 1000

compare =  function(las1, las2) { ... }
catalog_compare(multi_ctg, compare)
```

**Reply by KBodolai** [reply]
I was reading through the discussions, seeing if there's been much about what you just mentioned here:

> However pay attention to your data quality. It is unlikely that both dataset are perfectly aligned especially in Z.

I have some multitemporal data, and I want to make sure it's well aligned. It's over forests, so I'd like to use all the ground points together to derive the canopy height maps at both time points, but for that they need to be well aligned. 

I have done this manually in the past, but I'd like to do this programatically if possible. Have you dealt with this in the past? is there any resources you would recommend on the matter?

**Reply by wiesehahn** [reply]
I think you could for example select reference points at positions with no expected changes (e.g. along roads), calculate the average dtm difference at these positions, register one or the other pointcloud with this bias. This is more or less what they are doing in https://doi.org/10.1016/j.foreco.2023.121137.

I dont know how this is fone in other software but I also thought about if it would be possible to use some other registering methods than simply assuming the same vertical discrepancy in the entire pointcloud. Aybe providing positions with no expected difference and than use something like e.g. polynomial fit to adapt pointcloud z values (similar to how we would co-register images in XY dimensions).

**Reply by KBodolai** [reply]
I like the idea, unfortunately the data I'm looking to process is some of the [EBA lidar transects](https://www.nature.com/articles/s41597-023-02575-4#Sec2), which contains hundreds of acquisitions, of which a few will overlap. For some of them (e.g. transects 0076 and 0720) it may be impossible to find such invariant features. 

My initial idea was to try something similar for the whole DTM, given that overall, the lowest differences across the transects will be achieved when the heights are aligned (under some very mild assumptions). I also like the idea to do it more generally, where the displacement could be in any (x,y,z) dimension. Perhaps a similar thing can be made with shifting the DTM in X/Y and selecting the one which minimises the difference.

**Reply by Jean-Romain** [reply]
If you want me to develop custom tools for you specific need this now my full time job. See my website https://www.r-lidar.com/services. I'm currently integrating some point cloud registration algorithm in [lasR](https://github.com/r-lidar/lasR) as well as affine matrix transformation so I may already have a code base for your need.

I don't think `lidR` has what you need right now but I'm not sure to fully understand your requirements.

**Reply by KBodolai** [reply]
Hi @Jean-Romain, apologies if I came across as asking you to develop something specific, I was just wondering if you had come across this problem before.  I'm currently testing a couple of registration algorithms  that seem to perform well on this specific dataset, but it got me thinking about the problem more generally.

Great to know that you're working on this full time now though! I think I'll have to process quite a lot of ALS data in the near future, so I may get in touch soon :)

**Reply by Jean-Romain** [reply]
No offense. I was just informing you of my new status. If you are not interested it is not a problem. My software remain free and open source. If you are potentially interested I get a potential new a client. That's it.

I have never come across registration problem. That is why I never implemented anything. But you are lucky I will probably come into registration problem soon so I started developing tools. But the free and open source tools are coming "when it's ready" i.e. I don't know when :wink:

**Reply by KBodolai** [reply]
Excellent! Let me know if you can use a hand with that. I'm already writing a small benchmark of different approaches and it could at least serve as a small example in the documentation.
